Microsoft.Toolkit/Deferred/DeferredEventArgs.cs

```diff
@@ -0,0 +1,91 @@
+using System.Runtime.CompilerServices;
+
+namespace Microsoft.Toolkit.Deferred
+{
+    /// <summary>
+    /// <see cref="EventArgs"/> which can retrieve a <see cref="EventDeferral"/> in order to process data asynchronously before an <see cref="EventHandler"/> completes and returns to the calling control.
+    /// </summary>
+    public class DeferredEventArgs : EventArgs, IDisposable
+    {
+        /// <summary>
+        /// Gets a new <see cref="DeferredEventArgs"/> to use in cases where no <see cref="EventArgs"/> wish to be provided.
+        /// </summary>
+        public static new DeferredEventArgs Empty => new DeferredEventArgs();
+
+        private readonly object _eventDeferralLock = new object();
+
+        private EventDeferral? _eventDeferral;
+        private bool _disposed;
+
+        /// <summary>
+        /// Finalizes an instance of the <see cref="DeferredEventArgs"/> class.
+        /// </summary>
+        ~DeferredEventArgs() => Dispose(false);
+
+        /// <summary>
+        /// Returns an <see cref="EventDeferral"/> which can be completed when deferred event is ready to continue.
+        /// </summary>
+        /// <returns><see cref="EventDeferral"/> instance.</returns>
+        public EventDeferral GetDeferral()
+        {
+            lock (_eventDeferralLock)
+            {
+                return _eventDeferral ?? (_eventDeferral = new EventDeferral());
+            }
+        }
+
+        /// <summary>
+        /// DO NOT USE - This is a support method used by <see cref="EventHandlerExtensions"/>. It is public only for
+        /// additional usage within extensions for the UWP based TypedEventHandler extensions.
+        /// </summary>
+        /// <returns>Internal EventDeferral reference</returns>
+#if !NETSTANDARD1_4
+        [Browsable(false)]
+#endif
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        [Obsolete("This is an internal only method to be used by EventHandler extension classes, public callers should call GetDeferral() instead.")]
+        public EventDeferral? GetCurrentDeferralAndReset()
+        {
+            lock (_eventDeferralLock)
+            {
+                var eventDeferral = _eventDeferral;
+
+                _eventDeferral = null;
+
+                return eventDeferral;
+            }
+        }
+
+        /// <summary>
+        /// Recommended helper method pattern for <see cref="IDisposable"/>.
+        /// </summary>
+        /// <param name="disposing">Source of dispose request.</param>
+        protected virtual void Dispose(bool disposing)
+        {
+            if (_disposed)
+            {
+                return;
+            }
+
+            _eventDeferral?.Dispose();
+
+            _disposed = true;
+        }
+
+        /// <inheritdoc/>
+        public void Dispose()
+        {
+            // Dispose of unmanaged resources.
+            Dispose(true);
+
+            // Suppress finalization.
+            GC.SuppressFinalize(this);
+        }
```

Comment:
Technically this is suggested for types that also have unmanaged resources, to guarantee they're properly disposed no matter what. In this case it's still correct but possibly a bit overkill, we might want to simplify this in the future if we wanted. Just a note, the code itself is not wrong 🙂

Reply:
Yeah, I didn't want to do this stuff, but they're all raised as warnings/suggestions. So we'd have to turn those off or ignore them I guess... thoughts?

Reply:
The dispose method here is less related with the safe memory release and more with ease of use, as one can just use a `using` code block to release the deferral!

Reply:
How do we want consumers to use this type though? Because if we want them to always just use a `using` block for the deferral, we might as well also remove the destructor completely, which also reduces the overhead on the runtime/GC. It's true that when doing so, people could potentially cause bugs if they forget to dispose a deferral, but then again that should never happen so I'm not sure providing a safe fallback in that care is even really needed 🤔

Reply:
That's a very fair point! When I wrote this initially, I wanted to keep the same signature usage as one can find in UWP apps (so the whole `GetDeferral()` with  `Complete()`) but as that does require way more coding than just using an `IDisposable` inside a `using`.

The destructor is most likely an overkill, I initially add it to maintain the full `IDisposable` pattern, but I do not expect it to actually be called at any time!

Reply:
@michael-hawker following up on this, I do agree with @Sergio0694, removing the deconstructor and `GC.SuppressFinalize()` call makes sense as we actually do not care about it anyway for the `using` pattern (all we actually need is the `IDisposable` implementation and nothing else!)